tests/frontend/test_circuitspecs_chip2.py

```diff
@@ -0,0 +1,373 @@
+    """Dummy circuit used to instantiate
+    the abstract base class"""
+
+    modes = 8
+    remote = False
+    local = True
+    interactive = True
+    primitives = {"S2gate", "MeasureFock", "Rgate", "BSgate"}
+    decompositions = {"Interferometer": {}, "MZgate": {}}
+
+
+class TestChip0Compilation:
+    """Tests for compilation using the Chip0 circuit specification"""
+
+    def test_exact_template(self, tol):
+        """Test compilation works for the exact circuit"""
+        bb = blackbird.loads(Chip2Specs.circuit)
+        bb = bb(
+            squeezing_amplitude_0=0.43,
+            squeezing_amplitude_1=0.65,
+            squeezing_amplitude_2=0.543,
+            squeezing_amplitude_3=0.665,
+            phase_0=0,
+            phase_1=1,
+            phase_2=2,
+            phase_3=3,
+            phase_4=4,
+            phase_5=5,
+            phase_6=6,
+            phase_7=7,
+            phase_8=8,
+            phase_9=9,
+            phase_10=10,
+            phase_11=11,
+            final_phase_0=1.24,
+            final_phase_1=-0.54,
+            final_phase_2=4.12,
+            final_phase_3=0,
+            final_phase_4=1.24,
+            final_phase_5=-0.54,
+            final_phase_6=4.12,
+            final_phase_7=0,
+        )
+
+        expected = to_program(bb)
+        res = expected.compile("chip2")
+
+        assert program_equivalence(res, expected, atol=tol)
+
+    def test_not_all_modes_measured(self):
+        """Test exceptions raised if not all modes are measured"""
+        prog = sf.Program(8)
+        U = random_interferometer(4)
+
+        with prog.context as q:
+            ops.S2gate(0.5) | (q[0], q[4])
+            ops.S2gate(0.5) | (q[1], q[5])
+            ops.S2gate(0.5) | (q[2], q[6])
+            ops.S2gate(0.5) | (q[3], q[7])
+            ops.Interferometer(U) | (q[0], q[1], q[2], q[3])
+            ops.Interferometer(U) | (q[4], q[5], q[6], q[7])
+            ops.MeasureFock() | (q[0], q[1])
+
+        with pytest.raises(CircuitError, match="All modes must be measured"):
+            res = prog.compile("chip2")
+
+    def test_no_s2gates(self):
+        """Test exceptions raised if no S2gates are present"""
+        prog = sf.Program(8)
+        U = random_interferometer(4)
+
+        with prog.context as q:
+            ops.Interferometer(U) | (q[0], q[1], q[2], q[3])
+            ops.Interferometer(U) | (q[4], q[5], q[6], q[7])
+            ops.MeasureFock() | q
+
+        with pytest.raises(CircuitError, match="must start with four S2gates"):
+            res = prog.compile("chip2")
+
+    def test_incorrect_s2gates(self):
+        """Test exceptions raised if S2gates do not appear on correct modes"""
+        prog = sf.Program(8)
+        U = random_interferometer(4)
+
+        with prog.context as q:
+            ops.S2gate(0.5) | (q[0], q[1])
+            ops.S2gate(0.5) | (q[2], q[3])
+            ops.S2gate(0.5) | (q[4], q[5])
+            ops.S2gate(0.5) | (q[4], q[6])
```

Comment:
Might be better to test two things:
i) s2gates do not appear multiple times for same mode (this test)
ii) s2gates do not cross over between signal and idler modes (perhaps what this test's docstring would suggest)

Reply:
Yep I agree. This will need some thinking, I'm not sure the circuit spec compile method currently does this correctly.

The logic works for `chip0` I think by chance due to the small number of modes (4 modes, 2 S2gates), but I think the logic is incorrect, and doesn't scale for anything larger than 4 modes.

Worth fixing here, or making an issue? Since it likely affects chip0 as well.

Reply:
Let's start with the second suggested test, then fix things in the circuitspec if the test is failing

Reply:
Split this into two tests.